### ISSUE TITLE
Get stuck in load a batch 

### ISSUE THREAD

**Issue by liyunlongaaa** [wenet-e2e/wespeaker#114]
Hi, dear.

I am following the settings of the original script, using the shards data type, I want to know,  is your experiment using SSD? Is it because of the fuck mechanical hard disk that it has been stuck in load a batch? It is not very slow, but it is always stuck there forever.

I have a record about it

`

        print("before get a batch")
        for i, batch in enumerate(dataloader):
            print("get a batch")
            utts = batch['key']
            targets = batch['label']
            features = batch['feat']
`
![image](https://user-images.githubusercontent.com/49556860/204724276-5437eba8-d6d4-411c-a5ce-e7030876652f.png)

after, ... of, there is no after.

**Comment by JiJiJiang** [comment]
This seems unnormal. Using mechanical hard disk could be a bit slower, but not always stuck there. Maybe you should check: 1. DDP (use DP to debug), 2. cuda/nccl/pytorch version. Good luck to you.

**Comment by liyunlongaaa** [comment]
Do you all use solid-state drives in your lab,？cry""""

**Comment by JiJiJiang** [comment]
> Do you all use SSD in your lab,？cry""""

We have tried both SSD and mechanical hard disk. The difference of training speed should be fine and reasonable. ^_^

**Comment by liyunlongaaa** [comment]
thank you a lot !！I am persuading my boss to replace me with an SSD, because it is too slow, then according to what you said, it should be my problem, and there should be no need to waste this money emmm.....   oh I have a studi question，what is the type of shards function? it can  speed up data reading ？

**Comment by JiJiJiang** [comment]
Yes, using the shards file can speed up IO reading. Reading massive small wave files in an on-the-fly manner can be very slow, thus we combine tens of hundreds of wave files into a bigger shard file. Check [our paper](https://arxiv.org/abs/2210.17016) for more details.

**Comment by JiJiJiang** [comment]
BTW, is your cpu amd or intel?

**Comment by liyunlongaaa** [comment]
> BTW, is your cpu amd or intel?

my CPU is Intel(R) Xeon(R) CPU E5-2680 v3 @ 2.50GHz  ，wow thank you for your concern

**Comment by liyunlongaaa** [comment]
> Yes, using the shards file can speed up IO reading. Reading massive small wave files in an on-the-fly manner can be very slow, thus we combine tens of hundreds of wave files into a bigger shard file. Check [our paper](https://arxiv.org/abs/2210.17016) for more details.

I see I see , thx !

**Comment by liyunlongaaa** [comment]
oh, I found my error is caused by num_works, I set it to 0，it can work normally(我也不知道为啥). But.....it is not goog method(治标不治本)，the speed load data is can not catch up with my 3090 GPU... I am looking for a method can set num_works > 0 and work well.

**Comment by JiJiJiang** [comment]
> oh, I found my error is caused by num_works, I set it to 0，it can work normally(我也不知道为啥). But.....it is not goog method(治标不治本)，the speed load data is can not catch up with my 3090 GPU... I am looking for a method can set num_works > 0 and work well.

You could set num_workers as 2, 4, 8 to see if it works. Besides, num_workers should not exceed the cpu core num.

**Comment by liyunlongaaa** [comment]
hello, I set smaller num_works and it worked fine today, but I couldn't yesterday. I don't know what's going on, something similar happens in https://zhuanlan.zhihu.com/p/366595260。
![image](https://user-images.githubusercontent.com/49556860/204952565-937ad629-a2a8-4ac7-886b-11ad8eea0763.png)

but I get a new question, I use 3 GPU, but always 1 or 2 GPU can work in 100 percent-util . and after a while , they will Back to zero together for a long time。
![image](https://user-images.githubusercontent.com/49556860/204953175-b1b71d48-5650-461f-bcc7-5a45e8d72cc3.png)

it is normal ? the train is slow...

**Comment by JiJiJiang** [comment]
Yes, it's normal, because your machine IO speed is slow and num_workers is small. Use SSD or more num_workers should be better.
It depends on what GPU and how many GPUs you use. 
2*A100: ~15min per epoch
2*2080ti: 40~50min per epoch

**Comment by liyunlongaaa** [comment]
thank you ! help me a lot~
